### PR TITLE
Expiry extraction + new text detector

### PR DESCRIPTION
**Pull request by xsl**

### CODE REVIEW

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt [hunk 1]
```diff
@@ -607,24 +622,27 @@ class CardScanActivity :
...
             config = ResultAggregatorConfig.Builder()
-                .withMaxTotalAggregationTime(if (enableNameExtraction) 15.seconds else 2.seconds)
+                .withMaxTotalAggregationTime(if (enableNameExtraction || enableExpiryExtraction) 15.seconds else 2.seconds)
                 .withDefaultMaxSavedFrames(0)
                 .build(),
             listener = this,
-            requiredPanAgreementCount = 5,
-            requiredNameAgreementCount = 3,
-            isNameExtractionEnabled = enableNameExtraction
+            requiredPanAgreementCount = 2,
```
Comment (awushensky) [inline cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt:629]: can we lower this to 2? I believe @dxaen was concerned about dropping this below 5 for accuracy purposes.

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt [hunk 2]
```diff
@@ -607,24 +622,27 @@ class CardScanActivity :
...
+            requiredNameAgreementCount = 2,
```
Comment (awushensky) [inline cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt:630]: keep in mind that this value is the value at which name extraction will stop. If we set it to 3, we will search until max time has elapsed or we have reached 3 agreements. In the event that max time has elapsed, we will return the name with 2 agreements (this is hard-coded in the result aggregator).

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt [hunk 3]
```diff
@@ -70,41 +80,85 @@ class OcrResultAggregator(
...
+        updateState: (PaymentCardOcrState) -> Unit
+    ) {
         val pan = result.pan
-        val numberCount = if (pan != null && interimResult.hasValidPan) {
+        val numberCount = if (pan != null && isValidPan(result.pan)) {
             startAggregationTimer()
             panResults.countResult(pan) // This must be last so numberCount is assigned.
         } else 0
 
         if (!isPanScanningComplete && requiredPanAgreementCount != null && numberCount >= requiredPanAgreementCount) {
```
Comment (awushensky) [inline cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt:126]:
```suggestion
        if (requiredPanAgreementCount != null && numberCount >= requiredPanAgreementCount) {
```

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt [hunk 4]
```diff
@@ -70,41 +80,85 @@ class OcrResultAggregator(
...
+    /**
+     * Updates internal counter for expiry, and associated states for finishing the aggregator
+     */
+    private suspend fun updateNameState(name: String?) {
         val nameCount = if (name?.isNotEmpty() == true) {
             nameResults.countResult(name)
         } else 0
 
-        if (!isNameFound && requiredPanAgreementCount != null && nameCount >= requiredPanAgreementCount) {
+        if (!isNameFound && requiredNameAgreementCount != null && nameCount >= requiredNameAgreementCount) {
```
Comment (awushensky) [inline cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt:146]:
```suggestion
        if (requiredNameAgreementCount != null && nameCount >= requiredNameAgreementCount) {
```

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt [hunk 5]
```diff
@@ -70,41 +80,85 @@ class OcrResultAggregator(
...
+     * Updates internal counter for expiry, and associated states for finishing the aggregator
+     */
+    private suspend fun updateExpiryState(expiry: ExpiryDetect.Expiry?) {
+        val expiryCount = if (expiry != null) {
+            expiryResults.countResult(expiry)
         } else {
-            interimResult to null
+            0
+        }
+        if (!isExpiryFound && requiredExpiryAgreementCount != null && expiryCount >= requiredExpiryAgreementCount) {
```
Comment (awushensky) [inline cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/result/OcrResultAggregator.kt:160]:
```suggestion
        if (requiredExpiryAgreementCount != null && expiryCount >= requiredExpiryAgreementCount) {
```

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt (hunk 2 above)
Comment (xsl) [reply cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt:630]: yep, understood

File: cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt (hunk 1 above)
Comment (xsl) [reply cardscan-ui/src/main/java/com/getbouncer/cardscan/ui/CardScanActivity.kt:629]: i'll ping @dxaen. 5 takes so long that even on my pixel 3, for some cards, we'll barely have time to run anything else within the 15 seconds

**Pull request merged**